manual/src/tutorials/parallelism.etex

```diff
@@ -0,0 +1,732 @@
+
+\subsection{s:par_systhread_interaction}{Interaction with systhreads}
+
+How do systhreads interact with domains? The systhreads created on a particular
+domain remain pinned to that domain. Only one systhread at a time is allowed to
+run OCaml code on a particular domain. However, systhreads belonging to a
+particular domain may run C library or system code in parallel. Systhreads
+belonging to different domains may execute in parallel.
+
+When using systhreads, the thread created for executing the computation given
+to "Domain.spawn" is also treated as a systhread. For example, the following
+program creates in total two domains (including the initial domain) with two
+systhreads each (including the initial systhread for each of the domains).
+
+\begin{verbatim}
+(* dom_thr.ml *)
+let m = Mutex.create ()
+let r = ref None (* protected by m *)
+
+let task () =
+  let my_thr_id = Thread.(id (self ())) in
+  let my_dom_id :> int = Domain.self () in
+  Mutex.lock m;
+  begin match !r with
+  | None ->
+      Printf.printf "Thread %d running on domain %d saw initial write\n%!"
+        my_thr_id my_dom_id
+  | Some their_thr_id ->
+      Printf.printf "Thread %d running on domain %d saw the write by thread %d\n%!"
+        my_thr_id my_dom_id their_thr_id;
+  end;
+  r := Some my_thr_id;
+  Mutex.unlock m
+
+let task' () =
+  let t = Thread.create task () in
+  task ();
+  Thread.join t
+
+let main () =
+  let d = Domain.spawn task' in
+  task' ();
+  Domain.join d
+
+let _ = main ()
+\end{verbatim}
+
+\begin{verbatim}
+$ ocamlopt -I +threads unix.cmxa threads.cmxa -o dom_thr.exe dom_thr.ml
+$ ./dom_thr.exe
+Thread 1 running on domain 1 saw initial write
+Thread 0 running on domain 0 saw the write by thread 1
+Thread 2 running on domain 1 saw the write by thread 0
+Thread 3 running on domain 0 saw the write by thread 2
+\end{verbatim}
+
+This program uses a shared reference cell protected by a mutex to communicate
+between the different systhreads running on two different domains. The
+systhread identifiers uniquely identify systhreads in the program. The initial
+domain gets the domain id and the thread id as 0. The newly spawned domain gets
+domain id as 1.
+
+\section{s:par_atomics}{Atomics}
+
+Mutex, condition variables and semaphores are used to implement blocking
+synchronisation between domains. For non-blocking synchronisation, OCaml
+provides \stdmoduleref{Atomic} variables. As the name suggests, non-blocking
+synchronisation does not provide mechanisms for suspending and waking up
+domains. On the other hand, primitives used in non-blocking synchronisation are
+often compiled to atomic read-modify-write primitives that the hardware
+provides. As an example, the following program increments a non-atomic counter
+and an atomic counter in parallel.
+
+\begin{caml_example*}{verbatim}
+(* incr.ml *)
+let twice_in_parallel f =
+  let d1 = Domain.spawn f in
+  let d2 = Domain.spawn f in
+  ignore @@ Domain.join d1;
+  ignore @@ Domain.join d2
```

Comment:
```suggestion
  Domain.join d1;
  Domain.join d2
```

`ignore` is useless here because `Domain.join` returns `unit`.

Reply:
Done.